client/src/modules/stock/lots/modals/schedule.modal.js

```diff
@@ -0,0 +1,129 @@
+angular.module('bhima.controllers')
+  .controller('LotsScheduleModalController', LotsScheduleModalController);
+
+// dependencies injections
+LotsScheduleModalController.$inject = [
+  'data', '$uibModalInstance', 'StockService', 'LotService',
+  'NotifyService', 'bhConstants', 'moment', '$translate',
+];
+
+function LotsScheduleModalController(data, Instance, Stock, Lots,
+  Notify, bhConstants, Moment, $translate) {
+  const vm = this;
+  vm.close = close;
+  vm.lotUuid = data.uuid; // The lot that invoked this modal
+  vm.DATE_FMT = bhConstants.dates.format;
+  vm.labelWidth = 100; // pixels
+  vm.monthWidth = 36; // pixels
+  vm.numMonths = 36; // Number of table columns for months
+  vm.chartWidth = vm.labelWidth + (vm.numMonths * vm.monthWidth); // pixels
+
+  const today = new Date();
+  vm.startDate = new Date(today.getFullYear(), today.getMonth(), 1);
+
+  function startup() {
+    Stock.lots.read(null, {
+      inventory_uuid : data.inventoryUuid,
+      depot_uuid : data.depotUuid,
+    })
+      .then((lots) => {
+        // Save info about the inventory article (lot[0] may have zero quantity)
+        vm.inventory_name = lots[0].text;
+        vm.projection_text = $translate.instant('LOTS_SCHEDULE.PROJECTION_BASIS', lots[0]);
+
+        // We need to eliminate any exhausted lots and any expired lots
+        // and then sort the remaining lots by expiration date
+        vm.lots = lots.filter(lot => lot.quantity > 0)
+          .filter(lot => Moment(new Date(lot.expiration_date)) >= Moment(today))
+          .sort(sortByExpirationDate);
+
+        // runningDate is the date the last lot ran out
+        // (Always start the first lot at the current date; ignore past)
+        let runningDate = new Date(today);
+
+        vm.lots.forEach(lot => {
+          // Process the lots and determine sequential start/end dates and other info
+          lot.start_date = new Date(runningDate);
+
+          // Compute when the lot runs out (based on adjusted start date)
+          lot.exhausted_date = Moment(lot.start_date).add(lot.quantity / lot.avg_consumption, 'months').toDate();
+
+          // Compute the end date for this lot
+          lot.end_date = new Date(lot.expiration_date);
+          if (lot.exhausted_date < lot.expiration_date) {
```

Comment:
This logic is backwards - if the `expiration_date` is before the `exhausted_date`, then we need to make the `exhausted_date` the `expiration_date`, since we cannot use expired stock.

However, I still think it is best to keep both around and show the `end_date` as the date you will run out, but with X quantity wasted.  The interpretation being that the `end_date` will be the last day you would consume at the current consumption schedule.

Reply:
![image](https://user-images.githubusercontent.com/896472/126329907-00d7c5bb-05e5-49d3-982b-d20324af0770.png)

Once we make that correction to the logic, we need to make sure the scenario above is possible.  I've just edited it to show a hypothetical scenario where the stock expires before it runs out.  The overlap is _unusable stock_.  While the institution won't be consuming it, it shows the number of months that they are missing out on, if only they better managed their stock.

Reply:
Made these fixes.  I had thought of the red unusable residual block so I was happy to do this.  I also did some adjustments on the spacing to make the dates work on the graphical display.